src/main/java/org/betonquest/betonquest/events/PickRandomEvent.java

```diff
@@ -70,53 +74,35 @@ public PickRandomEvent(final Instruction instruction) throws InstructionParseExc
 
     @Override
     protected Void execute(final Profile profile) throws QuestRuntimeException {
-        final List<RandomEvent> events = new ArrayList<>(this.events);
-        double total = 0;
-        // Calculate total amount of all "percentages" (so that it must not be 100)
-        for (final RandomEvent event : events) {
-            total += event.getChance().getDouble(profile);
+        final List<ResolvedRandomEvent> events = new ArrayList<>();
+        for (final RandomEvent randomEvent : this.events) {
+            events.add(new ResolvedRandomEvent(randomEvent, profile));
         }
-        //pick as many events as given with pick optional (or 1 if amount wasn't specified)
+        double total = events.stream().mapToDouble(ResolvedRandomEvent::chance).sum();
+
         int pick = this.amount == null ? 1 : this.amount.getInt(profile);
         while (pick > 0 && !events.isEmpty()) {
             pick--;
-            //choose a random number between 0 and the total amount of percentages
-            final double found = Math.random() * total;
-            double current = 0;
-            //go through all random events and pick the first one where the current sum is higher than the found random number
-            for (int i = 0; i < events.size(); i++) {
-                final RandomEvent event = events.get(i);
-                final double chance = event.getChance().getDouble(profile);
-                current += chance;
-                if (current >= found) {
-                    //run the event
-                    BetonQuest.event(profile, event.getIdentifier());
-                    //remove the event from the list so that it's not picked again
-                    events.remove(i);
-                    total -= chance;
+            double random = randomGenerator.nextDouble() * total;
+            for (final ResolvedRandomEvent event : events) {
+                random -= event.chance;
+                if (random < 0) {
+                    BetonQuest.event(profile, event.identifier);
+                    events.remove(event);
+                    total -= event.chance;
```

Comment:
It would be better to use an explicit `Iterator` here as one can call `.remove()` on it which is much better optimized (this is in conjunction to the ArrayList -> LinkedList comment):
```suggestion
            for (final Iterator<ResolvedRandomEvent> iterator = events.iterator(); iterator.hasNext(); ) {
                event = iterator.next();
                random -= event.chance;
                if (random < 0) {
                    BetonQuest.event(profile, event.identifier);
                    iterator.remove();
                    total -= event.chance;
```